lib/llm/src/kv_router/scheduler.rs

```diff
@@ -242,51 +242,60 @@ impl WorkerSelector for DefaultWorkerSelector {
+        // 1. Handle edge cases
+        if workers.endpoints.is_empty() {
+            // No endpoints at all
+            return Err(KvSchedulerError::NoEndpoints);
         }
 
-        if max_active == 0.0 {
+        // If block_size == 0 or isl_tokens == 0, we consider that
+        // no blocks are required. We'll still pick a worker if possible,
+        // but the overlap-based score effectively becomes 0.
+        let zero_tokens_or_blocks = (request.isl_tokens == 0) || (block_size == 0);
+        let block_size_f = block_size as f64;
+        let isl_tokens_f = request.isl_tokens as f64;
+
+        // 2. Find the maximum request_active_slots to normalize usage
+        let max_active = workers
+            .endpoints
+            .values()
+            .map(|ep| ep.data.request_active_slots)
+            .max()
+            .unwrap_or(0);
+
+        // If all endpoints have 0 active slots, we can either:
+        // - Return an error (like the original code), or
+        // - Still proceed with the logic (since we do have endpoints).
+        if max_active == 0 {
+            // Decide how you want to handle this scenario.
+            // We'll return an error as the original code does:
             return Err(KvSchedulerError::NoEndpoints);
         }
 
-        // make immutable
-        let worker_scores = worker_scores;
-        let max_active = max_active;
-
-        // Calculate logits for each worker
+        // 3. In a single pass, determine the best worker(s) based on the logit formula
         let mut best_logit = f64::NEG_INFINITY;
         let mut best_workers = Vec::new();
 
-        for (worker_id, ep) in workers.endpoints.iter() {
-            let worker_id = *worker_id;
-
-            // Get score or default to 0.0
-            let score = worker_scores.get(&worker_id).copied().unwrap_or(0.0);
+        for (worker_id, ep) in &workers.endpoints {
+            // Overlap score (defaults to 0 if missing)
+            let overlap_score = request.overlap.scores.get(worker_id).copied().unwrap_or(0) as f64;
 
-            // Calculate normalized metrics
-            assert!(ep.data.kv_total_blocks > 0);
-            let gpu_cache_usage = ep.data.kv_active_blocks as f64 / ep.data.kv_total_blocks as f64;
-            let normalized_active = if max_active > 0.0 {
-                ep.data.request_active_slots as f64 / max_active
-            } else {
+            // If we have zero tokens or block size, we treat the effective 'score' as 0
+            // because we can't allocate blocks or meaningfully normalize the overlap.
+            let score = if zero_tokens_or_blocks {
                 0.0
+            } else {
+                overlap_score * block_size_f / isl_tokens_f
             };
 
-            // Calculate logit using same formula as Python
+            // Avoid division by zero; skip or handle endpoints that have zero kv_total_blocks
+            if ep.data.kv_total_blocks == 0 {
+                continue;
```

Comment:
this is better. thanks.  In theory, at least in the original code,  ProcessedEndpoints does the filtering, so only endpoints with reported non-zero blocks get by the filter.

But you are correct in checking. Someone could easily come along and change the behavior of processed endpoints, so we can't assume it always does the right thing.

Reply:
ProcessedEndpoints does seem to be a shell of its former self, so this is a good check.

I had initially done a filter in ProcessedEndpoints so we could avoid looping over workers that are not schedulable until the next load update.

Reply:
Again, we can do better.  Unit tests on ProcessedEndpoints can guarantee it conforms to a validated policy and we can avoid downstream validation knowing that ProcessedEndpoints is well tested and behaves in a given manner.